File: trunk/proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/reports/DocumentIndexItemWithBalance.java
package com.jasonzqshen.familyaccounting.core.reports;

import java.util.ArrayList;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataIdentity;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataType;
import com.jasonzqshen.familyaccounting.core.transaction.HeadEntity;
import com.jasonzqshen.familyaccounting.core.transaction.ItemEntity;
import com.jasonzqshen.familyaccounting.core.transaction.MonthIdentity;
import com.jasonzqshen.familyaccounting.core.utils.CreditDebitIndicator;
import com.jasonzqshen.familyaccounting.core.utils.CurrencyAmount;

/**
 * 
 * @author Jason
 * 
 */
public class DocumentIndexItemWithBalance extends DocumentIndexItem {

	private final MasterDataType _type;

	private final CurrencyAmount _sum;

	/**
	 * 
	 * @param coreDriver
	 * @param id
	 */
	DocumentIndexItemWithBalance(CoreDriver coreDriver, MasterDataIdentity id,
			MasterDataType type) {
		super(coreDriver, id);

		_type = type;
		_sum = new CurrencyAmount();
	}

	@Override
	protected void addDoc(HeadEntity head) {
		super.addDoc(head);

		// add the balance items
		ItemEntity[] items = head.getItems();
		for (ItemEntity item : items) {
			_sum.addTo(getAmountFromItem(item));
		}
	}

	/**
	 * remove document
	 */
	protected void removeDoc(HeadEntity head) {
		super.removeDoc(head);

		// add the balance items
		ItemEntity[] items = head.getItems();
		for (ItemEntity item : items) {
			CurrencyAmount amount = getAmountFromItem(item);
			amount.negate();
			_sum.addTo(amount);
		}
	}

	/**
	 * get the amount from item
	 * 
	 * @param item
	 * @return
	 */
	private CurrencyAmount getAmountFromItem(ItemEntity item) {
		if (_type == MasterDataType.BUSINESS_AREA) {
			MasterDataIdentity business = item.getBusinessArea();
			if (this._id.equals(business)) {
				if (item.getCDIndicator() == CreditDebitIndicator.DEBIT) {
					return item.getAmount();
				} else {
					CurrencyAmount amount = item.getAmount();
					amount.negate();
					return amount;
				}
			}
		} else if (_type == MasterDataType.CUSTOMER) {
			MasterDataIdentity customer = item.getCustomer();
			if (this._id.equals(customer)) {
				if (item.getCDIndicator() == CreditDebitIndicator.DEBIT) {
					return item.getAmount();
				} else {
					CurrencyAmount amount = item.getAmount();
					amount.negate();
					return amount;
				}
			}
		} else if (_type == MasterDataType.VENDOR) {
			MasterDataIdentity vendor = item.getVendor();
			if (this._id.equals(vendor)) {
				if (item.getCDIndicator() == CreditDebitIndicator.DEBIT) {
					return item.getAmount();
				} else {
					CurrencyAmount amount = item.getAmount();
					amount.negate();
					return amount;
				}
			}
		}

		return new CurrencyAmount();
	}

	/**
	 * get amount sum
	 * 
	 * @return
	 */
	public CurrencyAmount getAmountSum() {
		return new CurrencyAmount(_sum);
	}

	/**
	 * get amount
	 * 
	 * @param startId
	 * @param endId
	 * @return
	 */
	public CurrencyAmount getAmount(MonthIdentity startId, MonthIdentity endId) {
		CurrencyAmount amount = new CurrencyAmount();
		ArrayList<HeadEntity> docs = this.getEntities(startId, endId);
		for (HeadEntity head : docs) {
			ItemEntity[] items = head.getItems();
			for (ItemEntity item : items) {
				amount.addTo(getAmountFromItem(item));
			}
		}

		return amount;
	}
}

File: branches/V1.0/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Reports/DocumentIndexItemWithBalance.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Transaction;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.Reports
{
    public class DocumentIndexItemWithBalance : DocumentIndexItem
    {
        private readonly MasterDataType _type;

        private readonly CurrencyAmount _sum;

        /// <summary>
        /// constructor
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="id"></param>
        /// <param name="type"></param>
        internal DocumentIndexItemWithBalance(CoreDriver coreDriver, MasterDataIdentity id,
                MasterDataType type)
            : base(coreDriver, id)
        {
            _type = type;
            _sum = new CurrencyAmount();
        }

        /// <summary>
        /// Add document
        /// </summary>
        /// <param name="head"></param>
        internal override void addDoc(HeadEntity head)
        {
            base.addDoc(head);

            // add the balance items
            List<ItemEntity> items = head.Items;
            foreach (ItemEntity item in items)
            {
                _sum.AddTo(getAmountFromItem(item));
            }
        }

        /// <summary>
        /// remove document
        /// </summary>
        /// <param name="head"></param>
        internal override void removeDoc(HeadEntity head)
        {
            base.removeDoc(head);

            // add the balance items
            List<ItemEntity> items = head.Items;
            foreach (ItemEntity item in items)
            {
                CurrencyAmount amount = getAmountFromItem(item);
                amount.Negate();
                _sum.AddTo(amount);
            }
        }

        /// <summary>
        /// get the amount from item
        /// </summary>
        /// <param name="item"></param>
        /// <returns></returns>
        private CurrencyAmount getAmountFromItem(ItemEntity item)
        {
            if (_type == MasterDataType.BUSINESS_AREA)
            {
                MasterDataIdentity business = item.BusinessArea;
                if (this._id.Equals(business))
                {
                    if (item.CdIndicator == CreditDebitIndicator.DEBIT)
                    {
                        return item.Amount;
                    }
                    else
                    {
                        CurrencyAmount amount = item.Amount;
                        amount.Negate();
                        return amount;
                    }
                }
            }
            else if (_type == MasterDataType.CUSTOMER)
            {
                MasterDataIdentity customer = item.Customer;
                if (this._id.Equals(customer))
                {
                    if (item.CdIndicator == CreditDebitIndicator.DEBIT)
                    {
                        return item.Amount;
                    }
                    else
                    {
                        CurrencyAmount amount = item.Amount;
                        amount.Negate();
                        return amount;
                    }
                }
            }
            else if (_type == MasterDataType.VENDOR)
            {
                MasterDataIdentity vendor = item.Vendor;
                if (this._id.Equals(vendor))
                {
                    if (item.CdIndicator == CreditDebitIndicator.DEBIT)
                    {
                        return item.Amount;
                    }
                    else
                    {
                        CurrencyAmount amount = item.Amount;
                        amount.Negate();
                        return amount;
                    }
                }
            }

            return new CurrencyAmount();
        }

        /// <summary>
        /// get amount sum
        /// </summary>
        /// <returns></returns>
        public CurrencyAmount getAmountSum()
        {
            return new CurrencyAmount(_sum);
        }

        /// <summary>
        /// get amount
        /// </summary>
        /// <param name="startId">start month identity</param>
        /// <param name="endId">end month identity</param>
        /// <returns></returns>
        public CurrencyAmount getAmount(MonthIdentity startId, MonthIdentity endId)
        {
            CurrencyAmount amount = new CurrencyAmount();
            List<HeadEntity> docs = this.getEntities(startId, endId);
            foreach (HeadEntity head in docs)
            {
                List<ItemEntity> items = head.Items;
                foreach (ItemEntity item in items)
                {
                    amount.AddTo(getAmountFromItem(item));
                }
            }

            return amount;
        }
    }
}
